sonia_hardware_states/src/sonia_hardware_states/activate_io.py

```diff
@@ -1,59 +1,69 @@
+        -- action               uint8       see ActuatorDoAction message
 
         <= continue                         Activation of the element successful
-        <= failed                           Failed to activate the element (service issue)
+        <= failed                           Failed to activate the element (Most likely due to kill switch being on)
+        <= timed out                        Activation timed out
 
     '''
 
-    def __init__(self, element, side, timeout=10):
-        super(activate_io, self).__init__(outcomes=['continue', 'failed'])
+    def __init__(self, element, side, action, timeout=8):
+        super(activate_io, self).__init__(outcomes=['continue', 'failed', "timed out"])
         
-        self.start_time = None
-        self.do_action = None
-        self.param_side = side
-        self.param_element = element
+        self.time_start = None
+        self.element = element
+        self.side = side
+        self.action = action
         self.timeout = timeout
+        self.response = ""
 
-        self.action = ActuatorDoActionSrvRequest()
-        self.action_side = {'1': self.action.SIDE_PORT, '2': self.action.SIDE_STARBOARD}
-        self.action_element = {'1': self.action.ELEMENT_TORPEDO, '2': self.action.ELEMENT_DROPPER, '3': self.action.ELEMENT_ARM}
+        self.action_pub = rospy.Publisher('/provider_actuators/do_action_to_actuators', ActuatorDoAction, queue_size=2)
+        self.timeout_pub = rospy.Publisher('/sonia_behaviors/timeout', MissionTimer, queue_size=5)
 
     def on_enter(self, userdata):
-        try:
-            rospy.wait_for_service('/proc_actuators/cm_action_srv', self.timeout)
-            self.do_action = rospy.ServiceProxy('/proc_actuators/cm_action_srv', ActuatorDoActionSrv)
-            # rospy.wait_for_service('/provider_actuators/do_action_srv', self.timeout)
-            # self.do_action = rospy.ServiceProxy('/provider_actuators/do_action_srv', ActuatorDoActionSrv)
-        except rospy.ServiceException as exc:
-            rospy.loginfo('Service is not available : ' + str(exc))
-            return 'failed'
-
-        try:
-            self.do_action(self.action_element[str(int(self.param_element))], self.action_side[str(int(self.param_side))], self.action.ACTION_DROPPER_LAUNCH)
-        except rospy.ServiceException as exc:
-            rospy.loginfo('Service did not process request : ' + str(exc))
-            return 'failed'
+        self.action_sub = rospy.Subscriber('/provider_actuators/do_action_from_actuators', ActuatorSendReply, self.reply)
+        self.action_pub.publish(ActuatorDoAction(self.element, self.side, self.action))
+        self.time_start = time()
+        self.timeout_pub.publish(missionTimerFunc(f"activate_io (element{self.element}, side{self.side})", self.timeout, str(self.time_start), 1))
+        rospy.loginfo(f'Action : {int(self.element)} is launch')
+
+    def reply(self, data):
```

Comment:
Should be reply_cb just for clarification that it's a callback

Reply:
done